src/java/org/apache/cassandra/db/monitoring/SaiSlowLog.java

```diff
@@ -0,0 +1,193 @@
+
+import java.beans.ConstructorProperties;
+import java.util.Comparator;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Objects;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.index.sai.QueryContext;
+import org.apache.cassandra.tracing.Tracing;
+
+public class SaiSlowLog
+{
+    private static final Logger logger = LoggerFactory.getLogger(SaiSlowLog.class);
+
+    private final AtomicInteger numSlowestQueries = new AtomicInteger();
+    // a structure that keeps the N slowest queries - don't use getter/setter so that it's not exposed over JMX
+    private final SaiSlowestQueriesQueue slowestQueries;
+    // KATE: make this false by default and decide on yaml property, -D flag, or JMX to enable; hook setEnabled() to change
+    private final AtomicBoolean enabled = new AtomicBoolean(true);
+
+    static final Comparator<SlowSaiQuery> SLOW_SAI_QUERY_COMPARATOR = (SlowSaiQuery o1, SlowSaiQuery o2) -> Long.compare(o2.getDuration(), o1.getDuration());
+
+    @VisibleForTesting
+    static SaiSlowLog instance = new SaiSlowLog();
+
+    public SaiSlowLog()
+    {
+        numSlowestQueries.set(DatabaseDescriptor.getSaiSlowLogNumSlowestQueries());
+        slowestQueries = new SaiSlowestQueriesQueue(numSlowestQueries.get());
+    }
+
+    public boolean isEnabled()
+    {
+        return enabled.get();
+    }
+
+    public synchronized void setEnabled(boolean newValue)
+    {
+        boolean oldValue = this.enabled.get();
+
+        if (oldValue != newValue)
+        {
+            this.enabled.set(newValue);
+            logger.info("SAI Slow Log is now {}", newValue ? "enabled" : "disabled");
+        }
+    }
+
+    // KATE below two will be used to set and get the number of slowest queries; make it configurable through property/nodetool?
+    public int getNumSlowestQueries()
+    {
+        return numSlowestQueries.get();
+    }
+
+    public void setNumSlowestQueries(int numSlowestQueries)
+    {
+        int oldValue = this.numSlowestQueries.get();
+        if (oldValue != numSlowestQueries)
+        {
+            if (numSlowestQueries < 1)
+            {
+                throw new IllegalArgumentException("sai_slow_log_num_slowest_queries must be > 0");
+            }
+            this.numSlowestQueries.set(numSlowestQueries);
+            resizeMinMaxHeap();
+            logger.info("sai_slow_log_num_slowest_queries set to {} (was {}).", numSlowestQueries, oldValue);
+        }
+    }
+
+    private void resizeMinMaxHeap()
+    {
+        slowestQueries.resize(numSlowestQueries.get());
+    }
+
+    public List<SlowSaiQuery> retrieveRecentSlowestCqlQueries()
+    {
+        List<SlowSaiQuery> queries = slowestQueries.getAndReset();
```

Comment:
why do we reset the queue ? what's the use case ?

this API will be used from Autobot, with a few users (maybe 10 people on a incident call) who are looking at the problem
if the first gets the list and then the list is lost then it is enough to refresh the page and you would have lost this precious information about the system

Reply:
Reset is needed in order to get _most recent_ slow queries. 
But I see your point, so I would suggest we have both - `getAndReset` and only `get. ` WDYT?
Otherwise, we get stuck with old slow queries until next restart. During debugging, we need recent data.